models/biocro/R/run.biocro.R

```diff
@@ -51,111 +44,177 @@ run.biocro <- function(lat, lon, metpath, soil.nc = NULL, config = config, coppi
       day1 <- NULL
       dayn <- NULL
     }
+    if (!is.null(day1) && day1 > lubridate::yday(starti)) {
+      lubridate::yday(starti) <- day1
+    }
+    if (!is.null(dayn) && dayn < lubridate::yday(starti)) {
+      lubridate::yday(plant_date) <- dayn
+    }
+    WetDat <- WetDat[WetDat$doy >= lubridate::yday(starti) & WetDat$doy <= lubridate::yday(endi), ]
 
-    # BLETCHEROUS HACK: BioCro 0.94 starts the run by subsetting weather data
-    # to day1:dayn, but it assumes the data start on DOY 1 and contain
-    # (yearlength*(24/timestep)) lines. This means that in practice, day1 and
-    # dayn are treated as "day of file" not "day of year".
-    # BioCro *does* handle DOY correctly downstream of the subsetting, so here
-    # we check if the current BioCro has fixed this assumption.
-    # If not, rescale day1 and dayn to be relative to the start of the input.
-    #   Scaling is derived by inverting Biocro's day->index equations.
-    biocro_checks_doy <- tryCatch(
-      {m <- BioCro::BioGro(WetDat = matrix(c(0,10,0,0,0,0,0,0),nrow = 1),
-                   day1 = 10, dayn = 10, timestep = 24);
-      class(m) == "BioGro"},
-      error = function(e){FALSE})
-    if (!biocro_checks_doy && min(WetDat[,"doy"])>1) {
-      if (!is.null(day1)){
-        # Biocro calculates line number as `indes1 <- (day1 - 1) * 24`
-        indes1 <- Position(function(x)x==day1, WetDat[,"doy"])
-        day1 <- indes1/24 + 1
+    HarvestedYield <- 0
+
+    if (utils::packageVersion('BioCro') >= 1.0) {
+
+      if ("SolarR" %in% names(WetDat)) {
+        WetDat <- dplyr::rename(WetDat, solar = "SolarR")
       }
-      if (!is.null(dayn)){
-        # Biocro calculates line number as `indesn <- (dayn) * 24`
-        indesn <- Position(function(x)x==dayn, WetDat[,"doy"], right = TRUE)
-        dayn <- indesn/24
+      if ("WS" %in% names(WetDat)) {
+        WetDat <- dplyr::rename(WetDat, windspeed = "WS")
       }
-    }
-    
-    HarvestedYield <- 0
-    if (genus == "Saccharum") {
-      tmp.result <- BioCro::caneGro(WetDat = WetDat, lat = lat, soilControl = l2n(config$pft$soilControl))
-      # Addin Rhizome an Grain to avoid error in subsequent script processing results
-      tmp.result$Rhizome <- 0
-      tmp.result$Grain <- 0
-    } else if (genus == "Salix") {
+
       if (i == 1) {
-        iplant <- config$pft$iPlantControl
-      } else {
-        iplant$iRhizome <- data.table::last(tmp.result$Rhizome)
-        iplant$iRoot <- data.table::last(tmp.result$Root)
-        iplant$iStem <- data.table::last(tmp.result$Stem)
-        
-        if ((i - 1)%%coppice.interval == 0) {
+        initial_values <- config$pft$initial_values
+      }
+
+      if (i > 1) { # TODO HarvestedYield is never used and coppice not applicable to all crops. Rethink?
+        if ((i - 1) %% coppice.interval == 0) {
```

Comment:
@justinmcgrath this is logic for handling perennial crops like willow ... I implemented it in PEcAn to generalize the original approach (https://github.com/ebimodeling/biocro/blob/master/inst/extdata/willowGrotestv0.912.R) in which coppice cycles for a specific analysis are written into a script. 

I always thought this type of logic should go in the model. Does the rewrite do anything to support perennial crops, or could this logic be incorporated into the model itself?

Reply:
@justinmcgrath  great I see it here: https://github.com/ebimodeling/biocro-dev/issues/124

>  but doing that will require Pecan to examine the solution to determine how much was removed

Harvested biomass is certainly an output of interest, not only to PEcAn but to any user of BioCro for crop simulation - wouldn't it make sense for BioCro to keep track of what was removed as a separate variable 'HarvestedYield' (as I originally implemented here since it doesn't currently do so?)?

Reply:
Yes, that can be saved.